using_restricted_array.rb

```diff
@@ -52,34 +52,60 @@ def find_largest(array, length)
   max = array[0]
   until i == length
     if max < array[i]
-      array[i] = max
-    end
-    return max
-  end
+      max = array[i]
+    end #if loop
+    i += 1
+  end #until loop
+  return max
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  i =1
+  i = 1
   min = array[0]
   until i == length
     if min > array[i]
-      array[i] = min
-    end
-    return min
-  end
+      min = array[i]
+    end #if
+    i += 1
+  end #until
+  return min
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  min = 0
+  max = length - 1
+
+  until min >= max
+    temp = array[min]
+    array[min] = array[max]
+    array[max] = temp
+    min += 1
+    max -= 1
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  min = 0
```

Comment:
Amy, please take a look at the updated binary_search slide. Here's the consideration I was missing in the algorithm on the initial slide that this code is based on: (1) what if there is only one element in the array and it is the value_to_find? The current code will return false. (To fix this, add an additional if statement after the loop ending on line 107, to check if array[low] is the value_to_find.) (2) Consider there are two elements in the array and the value_to_find is the second element in the array. min = 0, max=1, mid=0. Value is not at index 0. So, min = mid (still 0). On re-entering the loop, mid is still 0. So we have an infinite loop. (To fix this, update max to be mid -1 on line 102, and update line 105 to be min = mid+1.)